src/js/common/schemaform/ArrayField.jsx

```diff
@@ -0,0 +1,324 @@
+import React from 'react';
+import _ from 'lodash/fp';
+import Scroll from 'react-scroll';
+import { focusElement } from '../utils/helpers';
+
+import {
+  retrieveSchema,
+  toIdSchema,
+  getDefaultFormState,
+  deepEquals
+} from 'react-jsonschema-form/lib/utils';
+
+import { errorSchemaIsValid } from './helpers';
+
+const Element = Scroll.Element;
+const scroller = Scroll.scroller;
+
+const scrollToFirstError = () => {
+  setTimeout(() => {
+    const errorEl = document.querySelector('.usa-input-error, .input-error-date');
+    if (errorEl) {
+      const position = errorEl.getBoundingClientRect().top + document.body.scrollTop;
+      Scroll.animateScroll.scrollTo(position - 10, {
+        duration: 500,
+        delay: 0,
+        smooth: true
+      });
+      focusElement(errorEl);
+    }
+  }, 100);
+};
+
+/* Non-review growable table (array) field */
+export default class ArrayField extends React.Component {
+  constructor(props) {
+    super(props);
+
+    /*
+     * We always want to show one empty item in the array, so add that if necessary
+     *
+     * It will not get saved unless a user changes data in it
+     */
+    const formData = Array.isArray(props.formData) ? props.formData : null;
+    let items;
+    if (!formData || formData.length === 0) {
+      items = [getDefaultFormState(props.schema.items, null, props.registry.definitions) || {}];
+    } else {
+      items = getDefaultFormState(props.schema, formData, props.registry.definitions) || [];
+    }
+
+    /*
+     * We're keeping the editing state in local state because it's easier to manage and
+     * doesn't need to persist from page to page
+     */
+    this.state = {
+      items,
+      editing: items.map(() => false)
+    };
+
+    this.onItemChange = this.onItemChange.bind(this);
+    this.handleAdd = this.handleAdd.bind(this);
+    this.handleEdit = this.handleEdit.bind(this);
+    this.handleUpdate = this.handleUpdate.bind(this);
+    this.handleRemove = this.handleRemove.bind(this);
+    this.onItemBlur = this.onItemBlur.bind(this);
+    this.scrollToTop = this.scrollToTop.bind(this);
+    this.scrollToRow = this.scrollToRow.bind(this);
+  }
+
+  /*
+   * Based on the default react-jsonschema-form optimizations, except it handles state/props syncing
+   * with formData
+   */
+  shouldComponentUpdate(nextProps, nextState) {
+    const propsWithoutDataUnchanged = deepEquals(_.omit('formData', this.props), _.omit('formData', nextProps));
+    const stateUnchanged = deepEquals(this.state, nextState);
+
+    // Because there's local state for the form data, we sometimes see form data updated in props that is already
+    // in state and we don't need to update
+    if (propsWithoutDataUnchanged && stateUnchanged && deepEquals(nextProps.formData, nextState.items)) {
```

Comment:
Yeah, there are two optimizations here. One is to use local state, which was in the library's implementation of ArrayField, so I carried it over. The second is what I added to make sure we don't unnecessarily update when the local state and the data from props are the same.

It's probably worth removing the local state and seeing if the performance impact is noticeable. I agree that it'd be better not to have it, I just figured the authors of the library added it for a reason.